Model/Api/OAuthRedirect.php

```diff
@@ -35,37 +46,198 @@ class OAuthRedirect implements OAuthRedirectInterface
+     * @param CustomerSession                  $customerSession
+     * @param CustomerInterfaceFactory         $customerInterfaceFactory
+     * @param CustomerFactory                  $customerFactory
+     * @param Url                              $url
+     * @param Bugsnag                          $bugsnag
      */
     public function __construct(
         Response $response,
-        DeciderHelper $deciderHelper
+        DeciderHelper $deciderHelper,
+        SSOHelper $ssoHelper,
+        ExternalCustomerEntityRepository $externalCustomerEntityRepository,
+        CustomerRepository $customerRepository,
+        StoreManagerInterface $storeManager,
+        CustomerSession $customerSession,
+        CustomerInterfaceFactory $customerInterfaceFactory,
+        CustomerFactory $customerFactory,
+        Url $url,
+        Bugsnag $bugsnag
     ) {
         $this->response = $response;
         $this->deciderHelper = $deciderHelper;
+        $this->ssoHelper = $ssoHelper;
+        $this->externalCustomerEntityRepository = $externalCustomerEntityRepository;
+        $this->customerRepository = $customerRepository;
+        $this->storeManager = $storeManager;
+        $this->customerSession = $customerSession;
+        $this->customerInterfaceFactory = $customerInterfaceFactory;
+        $this->customerFactory = $customerFactory;
+        $this->url = $url;
+        $this->bugsnag = $bugsnag;
     }
 
     /**
      * Login with Bolt SSO and redirect
      *
      * @api
      *
-     * @param string $authorization_code
+     * @param string $code
+     * @param string $scope
+     * @param string $state
      *
      * @return void
      *
      * @throws NoSuchEntityException
+     * @throws WebapiException
      */
-    public function login($authorization_code = '')
+    public function login($code = '', $scope = '', $state = '')
     {
         if (!$this->deciderHelper->isBoltSSOEnabled()) {
             throw new NoSuchEntityException(__('Request does not match any route.'));
         }
 
-        $this->response->setHeader('Content-Type', 'application/json');
-        $this->response->setHttpResponseCode(500);
-        $this->response->setBody(json_encode(['message' => 'Not implemented']));
-        $this->response->sendResponse();
+        if ($code === '' || $scope === '' || $state === '') {
+            throw new WebapiException(__('Bad Request'), 0, WebapiException::HTTP_BAD_REQUEST);
+        }
+
+        $oauthConfiguration = $this->ssoHelper->getOAuthConfiguration();
+        $clientID = $oauthConfiguration['clientID'];
+        $clientSecret = $oauthConfiguration['clientSecret'];
+        $boltPublicKey = $oauthConfiguration['boltPublicKey'];
+
+        $token = $this->ssoHelper->exchangeToken($code, $scope, $clientID, $clientSecret);
+        if ($token === null) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
```

Comment:
Do we need more detailed error messages? The same questions about other places

Reply:
According to @nickfiacco-bolt we should not have detailed error messages

Reply:
yeah I want to avoid exposing any information about the internals of our SSO system as much as possible

Reply:
I just afraid it can be hard to debug something. We can save error in bugsnag but up to you guys

Reply:
that's a good idea. @adentong can you add bugsnags for these errors?